### ISSUE TITLE
VOC training slower with CUDNN

### ISSUE THREAD

**Issue by regiscely** [AlexeyAB/darknet#461]
The VOC training is almost 2 times slower if I compile with CUDNN=1 rather that CUDA only (GPU=1 & CUDNN=0)
The hardware is Power 9 (ppc64le architecture) with 4 V100 GPUs
My software stack is RHEL 7.5, CUDA 9.2.35 and cudnn 7.1.1 and drivers are at 396.06 level
GCC version is 4.8.5

**Comment by AlexeyAB** [comment]
* Does training slow down only when training on 1 GPU or when training on several GPUs `-gpus 0,1,2,3`?  https://github.com/AlexeyAB/darknet#how-to-train-with-multi-gpu

* Did you train with `random=1` in the cfg file? If you train with `random=1`, then did you see this message when network resizing goes during training `used slow CUDNN algo without Workspace!`?

* Do you use latest commit?

* Try to use mixed-precision `GPU=1` `CUDNN=1` and change these 2 lines in the Makefile: https://github.com/AlexeyAB/darknet/blob/cad4d1618fee74471d335314cb77070fee951a42/Makefile#L72-L73
to these:
```
COMMON+= -DCUDNN -DCUDNN_HALF
CFLAGS+= -DCUDNN -I/usr/local/cudnn/include -DCUDNN_HALF
```

**Comment by regiscely** [comment]
Yes I am using the lastest commit.
Here are my results:
**CUDA only** 
1 GPU (batches from 0 to 1000)=>it takes 35s
4 GPUs (batches from 1000 to 2000) => it takes 35s

**CUDA + CUDNN**
1 GPU (batches from 0 to 1000)=>it takes 89s
4 GPUs (batches from 1000 to 2000) => it takes 50s

Trainings were all performed with `random=1` and I see:
```
Resizing
512
 try to allocate workspace = 18874368 * sizeof(float),  CUDA allocate done!
```
When I am using DCUDNN_HALF flag, it is worse
1 GPU (batches from 0 to 1000)=>takes 121s duration

I am stuck here - Any ideas are welcome!

**Comment by regiscely** [comment]
hum - I think that my batch size and subdivisions parameter were not correct (default testing values) causing side effect on my training time

**Comment by AlexeyAB** [comment]
@regiscely Yes, you should use `batch=64 or more` and `subdivsion=8 or less`
The larger the batch size and the smaller the subdivision - the faster the training will be.

* Also I would recommend to base your cfg-file on `yolo.2.0.cfg`/`yolo-voc.2.0.cfg` instead of `yolo.cfg`/`yolo-voc.cfg`

* Also you can try to set this value 8 - More CPU-treads will be used for data loading: https://github.com/AlexeyAB/darknet/blob/df076653e00db69c6fb57869981f5196a8f55e70/src/detector.c#L92

* What batch and subdivision do you use now, and what time it tooks for CUDA + CUDNN + DCUDNN_HALF?

**Comment by regiscely** [comment]
Here are my settings:
```
batch=64
subdivisions=8
```
and to shortener the tests I have set `max_batches = 500` (I am focused on the performances right now not on the accuracy)
CUDA only => 601 seconds
CUDA + CUDNN => 493 seconds
CUDA + CUDNN + DCUDNN_HALF => 522 seconds - I was expecting better performance results there. I am investigating why it is slower than CUDA + CUDNN test

**Comment by AlexeyAB** [comment]
@regiscely The actual size of the batch = `batch / subdivision`, i.e. actual batch size now 64/8=**2**.

* So you can try to set `subdivision=1` (then actual batch size will be **64**)

* and if this does not speed up the training, then you can increase batch more, for example, `batch=128` or 256 (then actual batch size will be **128 or 256**)

It should increase speed.

**Comment by AlexeyAB** [comment]
@regiscely Also I fixed bug that saves weights for each iteration. It led to a slowdown in training.

Bug was between these 2 commits:
1. Fixed bug, that save weights-file for each iteration - committed 7 hours ago 
2. shortcut_layer resize for random=1 - committed 20 hours ago
